runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WindmillStateInternals.java

```diff
@@ -2005,79 +2025,64 @@ public Iterable<Entry<K, V>> read() {
-                          try {
-                            final K key = keyCoder.decode(entry.getKey().newInput());
-                            final Object structuralKey = keyCoder.structuralValue(key);
-                            KeyState keyState =
-                                keyStateMap.computeIfAbsent(structuralKey, k -> new KeyState(key));
-                            if (keyState.existence == KeyExistence.UNKNOWN_EXISTENCE) {
-                              keyState.existence = KeyExistence.KNOWN_EXIST;
-                            }
-                            return new AbstractMap.SimpleEntry<>(structuralKey, entry.getValue());
-                          } catch (IOException e) {
-                            throw new RuntimeException(e);
-                          }
-                        }),
-                    entry -> {
-                      KeyState keyState = keyStateMap.get(entry.getKey());
-                      return keyState.existence != KeyExistence.KNOWN_NONEXISTENT
-                          && !(keyState.existence == KeyExistence.KNOWN_EXIST
-                              && keyState.valuesCached);
-                    });
 
             if (entries instanceof Weighted) {
               // This is a known amount of data, cache them all.
-              filteredEntries.forEach(
+              entries.forEach(
                   entry -> {
-                    final Object structuralKey = entry.getKey();
-                    KeyState keyState = keyStateMap.get(structuralKey);
-                    keyState.existence = KeyExistence.KNOWN_EXIST;
-                    keyState.values.extendWith(entry.getValue());
-                    // We can't set keyState.valuesCached to true here, because there may be more
-                    // paginated values that should not be filtered out in filteredEntries.
+                    try {
+                      final K key = keyCoder.decode(entry.getKey().newInput());
+                      final Object structuralKey = keyCoder.structuralValue(key);
+                      KeyState keyState =
+                          keyStateMap.computeIfAbsent(structuralKey, k -> new KeyState(key));
+                      // Ignore any key from windmill that has been marked pending deletion, and any
+                      // key that is fully cached.
+                      if (keyState.existence == KeyExistence.KNOWN_NONEXISTENT
+                          || (keyState.existence == KeyExistence.KNOWN_EXIST
+                              && keyState.valuesCached)) return;
+                      // Or else cache contents from windmill.
+                      keyState.existence = KeyExistence.KNOWN_EXIST;
+                      keyState.values.extendWith(entry.getValue());
+                      // We can't set keyState.valuesCached to true here, because there may be more
+                      // paginated values that should not be filtered out in above if statement.
```

Comment:
this isn't a paginated read, can we set valuesCached?
Or shoudl comment be updated that there may be additional entries for the key.